### PR TITLE
Update troubleshooting with granular debug information for windows an…

### PR DESCRIPTION
**Pull request by jayunit100**
Some improvements on how to debug hairy bootstrap issues.

### CODE REVIEW

File: docs/book/src/user/troubleshooting.md [hunk 1]
```diff
@@ -38,6 +48,25 @@ For convenience, here is an example one-liner to do this post installation
 kubectl get nodes --no-headers -l '!node-role.kubernetes.io/master' -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}' | xargs -I{} kubectl label node {} node-role.kubernetes.io/worker=''
 ```                  
 
+## Troubleshooting Windows Node Bootstrapping Errors
+
+- Check for errors in the bootstrapper you are using, such as `C:\Program Files\Cloudbase Solutions\Cloudbase-Init\log`.
+- Be sure to look at `DEBUG` logs, as you might have a fatal error hidden in these log messages.  Possible bootstrap errors are:
+  - The windows node is missing some Binaries which you expected to have in the image on boot.
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:55]: nit
```suggestion
  - The windows node is missing some binaries which you expected to have in the image on boot.
```

File: docs/book/src/user/troubleshooting.md [hunk 2]
```diff
@@ -38,6 +48,25 @@ For convenience, here is an example one-liner to do this post installation
...
+
+
+## Getting the bootstrap data manually to inspect it
+
+Ultimately, if a VM isnt joining/bootstrapping into the cluster, you'll probably just want to see wether or not it has the bootstrap data.  This is the most direct way to determine where in the chain of command your bootstrap is failing.
+
+- If your VM doesnt have the right bootstrap data, its because of a bug in the provider's ability to attach bootstrap data.
+- If your VM does have the right bootstrap data, then theres a bug in the way you are configuring its startup logic, and this is most often the case.
+
+Irrespective of OS, on any Vsphere VM, for you can confirm the bootstrap information in the `Configuration Parameters` section of your VSphere machine settings, which are available from the `VM Hardware -> Edit Settings -> Advanced -> Configuration Parameters` section, and then read them by copying them and running `base64 --decode` on their contents.
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:66]: Is this entire troubleshooting section only relevant to vSphere Windows nodes? We might want to make that really explicit everywhere it's the case if so.

File: docs/book/src/user/troubleshooting.md [hunk 3]
```diff
@@ -38,6 +48,25 @@ For convenience, here is an example one-liner to do this post installation
...
+## Getting the bootstrap data manually to inspect it
+
+Ultimately, if a VM isnt joining/bootstrapping into the cluster, you'll probably just want to see wether or not it has the bootstrap data.  This is the most direct way to determine where in the chain of command your bootstrap is failing.
+
+- If your VM doesnt have the right bootstrap data, its because of a bug in the provider's ability to attach bootstrap data.
+- If your VM does have the right bootstrap data, then theres a bug in the way you are configuring its startup logic, and this is most often the case.
+
+Irrespective of OS, on any Vsphere VM, for you can confirm the bootstrap information in the `Configuration Parameters` section of your VSphere machine settings, which are available from the `VM Hardware -> Edit Settings -> Advanced -> Configuration Parameters` section, and then read them by copying them and running `base64 --decode` on their contents.   
+
+Other clouds will have similar methodologies for inspecting the user data attached to a VM.
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:68]: Maybe make the vSphere issue a subsection and open an issue to add the information to troubleshoot this issue for other cloud providers ?

File: docs/book/src/user/troubleshooting.md (hunk 1 above)
Comment (killianmuldoon) [reply docs/book/src/user/troubleshooting.md:55]: nit 
```suggestion
  - The Windows node is missing some binaries which you expected to have in the image on boot.
```

File: docs/book/src/user/troubleshooting.md [hunk 4]
```diff
@@ -38,6 +48,25 @@ For convenience, here is an example one-liner to do this post installation
 kubectl get nodes --no-headers -l '!node-role.kubernetes.io/master' -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}' | xargs -I{} kubectl label node {} node-role.kubernetes.io/worker=''
 ```                  
 
+## Troubleshooting Windows Node Bootstrapping Errors
+
+- Check for errors in the bootstrapper you are using, such as `C:\Program Files\Cloudbase Solutions\Cloudbase-Init\log`.
+- Be sure to look at `DEBUG` logs, as you might have a fatal error hidden in these log messages.  Possible bootstrap errors are:
+  - The windows node is missing some Binaries which you expected to have in the image on boot.
+  - The windows node might be accidentally running powershell or choco commands during bootstrap that aren't installed.
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:56]: nit 
```suggestion
  - The Windows node might be accidentally running PowerShell or Chocalatey / choco commands during bootstrap that aren't installed. 
```

File: docs/book/src/user/troubleshooting.md [hunk 5]
```diff
@@ -38,6 +48,25 @@ For convenience, here is an example one-liner to do this post installation
 kubectl get nodes --no-headers -l '!node-role.kubernetes.io/master' -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}' | xargs -I{} kubectl label node {} node-role.kubernetes.io/worker=''
 ```                  
 
+## Troubleshooting Windows Node Bootstrapping Errors
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:51]: nit - don't use title case to make consistent with the rest of the text:
```suggestion
## Troubleshooting Windows node bootstrapping errors
```

File: docs/book/src/user/troubleshooting.md [hunk 6]
```diff
@@ -1,6 +1,16 @@
...
-## Node bootstrap failures when using CABPK with cloud-init
+## Interpretting token refresh messages
+
+Sometimes, you'll see that the bootstrapping secret is noted as not having been consumed yet, *even though* you correctly ran the bootstrapping phase on your node.  This happens because Cluster API didn't yet hear back from a VM that it expected to bootstrap and join the cluster.  In this case, you might see a recurrent:
+
+```
+Refreshing token until the infrastructure has a chance to consume it
+```
+
+message in the logs for your `capi-kubeadm-bootstrap-system` controller.
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:11]: There's some context missing here as to the meaning of this message or the follow on. All of the headings are on the same level so it looks like there's no logical flow for solving this problem once ID'd - maybe add links or change the section headings.

File: docs/book/src/user/troubleshooting.md [hunk 7]
```diff
@@ -1,6 +1,16 @@
 # Troubleshooting
 
-## Node bootstrap failures when using CABPK with cloud-init
+## Interpretting token refresh messages
+
+Sometimes, you'll see that the bootstrapping secret is noted as not having been consumed yet, *even though* you correctly ran the bootstrapping phase on your node.  This happens because Cluster API didn't yet hear back from a VM that it expected to bootstrap and join the cluster.  In this case, you might see a recurrent:
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:5]: Use of VM here is maybe too specific? Is this somehow limited to VMs vs BmMs?

File: docs/book/src/user/troubleshooting.md [hunk 8]
```diff
@@ -18,7 +28,7 @@ be troubleshooted systematically.
 1. If Node bootstrapping consistently fails and the kubeadm logs are not verbose enough, the `kubeadm` verbosity 
    can be increased via `KubeadmConfigSpec.Verbosity`.
 
-## Labeling nodes with reserved labels such as `node-role.kubernetes.io` fails with kubeadm error during bootstrap
+## Avoid mislabeling nodes with reserved labels such as `node-role.kubernetes.io` fails with kubeadm error during bootstrap
```
Comment (killianmuldoon) [inline docs/book/src/user/troubleshooting.md:31]: I would revert this - IMHO the original phrase is clearer.

**Review by killianmuldoon — CHANGES_REQUESTED** [review]
Looks good - I'm not able to test this on a Windows VM, but well structured solutions for common errors are definitely welcome. I think there's a couple of nits and issues with structure to clean up.

**Comment by fabriziopandini** [comment]
@jayunit100 do you have some time to address the comments above?

**Comment by enxebre** [comment]
Hey @jayunit100 are you still pursing this PR?

**Comment by vincepri** [comment]
Closing for now due to inactivity, please feel free to reopen once ready.

/close

**Comment by jayunit100** [comment]
Just saw , Sgtm sorry got busy w other stuff :)

**Pull request closed without merging**